com_github_smallru8_driver_tuntap_TunTap.cpp

```diff
@@ -0,0 +1,235 @@
+ * Method:    tuntap_get_hwaddr
+ * Signature: ()Ljava/lang/String;
+ */
+JNIEXPORT jstring JNICALL Java_com_github_smallru8_driver_tuntap_TunTap_tuntap_1get_1hwaddr
+  (JNIEnv *env, jobject obj){
+	return charTojstring(env,tuntap_get_hwaddr(dev));
+}
+
+/*
+ * Class:     com_github_smallru8_driver_tuntap_TunTap
+ * Method:    tuntap_set_hwaddr
+ * Signature: (Ljava/lang/String;)I
+ */
+JNIEXPORT jint JNICALL Java_com_github_smallru8_driver_tuntap_TunTap_tuntap_1set_1hwaddr
+  (JNIEnv *env, jobject obj, jstring hwaddr){
+	return tuntap_set_hwaddr(dev,jstringToChar(env,hwaddr));
+}
+
+/*
+ * Class:     com_github_smallru8_driver_tuntap_TunTap
+ * Method:    tuntap_set_descr
+ * Signature: (Ljava/lang/String;)I
+ */
+JNIEXPORT jint JNICALL Java_com_github_smallru8_driver_tuntap_TunTap_tuntap_1set_1descr
+  (JNIEnv *env, jobject obj, jstring descr){
+	return tuntap_set_descr(dev,jstringToChar(env,descr));
+}
+
+/*
+ * Class:     com_github_smallru8_driver_tuntap_TunTap
+ * Method:    tuntap_get_descr
+ * Signature: ()Ljava/lang/String;
+ */
+JNIEXPORT jstring JNICALL Java_com_github_smallru8_driver_tuntap_TunTap_tuntap_1get_1descr
+  (JNIEnv *env, jobject obj){
+	return charTojstring(env,tuntap_get_descr(dev));
+}
+
+/*
+ * Class:     com_github_smallru8_driver_tuntap_TunTap
+ * Method:    tuntap_up
+ * Signature: ()I
+ */
+JNIEXPORT jint JNICALL Java_com_github_smallru8_driver_tuntap_TunTap_tuntap_1up
+  (JNIEnv *env, jobject obj){
+	return tuntap_up(dev);
+}
+
+/*
+ * Class:     com_github_smallru8_driver_tuntap_TunTap
+ * Method:    tuntap_down
+ * Signature: ()I
+ */
+JNIEXPORT jint JNICALL Java_com_github_smallru8_driver_tuntap_TunTap_tuntap_1down
+  (JNIEnv *env, jobject obj){
+	return tuntap_down(dev);
+}
+
+/*
+ * Class:     com_github_smallru8_driver_tuntap_TunTap
+ * Method:    tuntap_get_mtu
+ * Signature: ()I
+ */
+JNIEXPORT jint JNICALL Java_com_github_smallru8_driver_tuntap_TunTap_tuntap_1get_1mtu
+  (JNIEnv *env, jobject obj){
+	return tuntap_get_mtu(dev);
+}
+
+/*
+ * Class:     com_github_smallru8_driver_tuntap_TunTap
+ * Method:    tuntap_set_mtu
+ * Signature: ()I
+ */
+JNIEXPORT jint JNICALL Java_com_github_smallru8_driver_tuntap_TunTap_tuntap_1set_1mtu
+  (JNIEnv *env, jobject obj,jint mtu){
+	return tuntap_set_mtu(dev,mtu);
+}
+
+/*
+ * Class:     com_github_smallru8_driver_tuntap_TunTap
+ * Method:    tuntap_set_ip
+ * Signature: (Ljava/lang/String;I)I
+ */
+JNIEXPORT jint JNICALL Java_com_github_smallru8_driver_tuntap_TunTap_tuntap_1set_1ip
+  (JNIEnv *env, jobject obj, jstring ipaddr, jint mask){
+	return tuntap_set_ip(dev,jstringToChar(env,ipaddr),mask);
+}
+
+/*
+ * Class:     com_github_smallru8_driver_tuntap_TunTap
+ * Method:    tuntap_read
+ * Signature: (I)[B
+ */
+JNIEXPORT jbyteArray JNICALL Java_com_github_smallru8_driver_tuntap_TunTap_tuntap_1read
+  (JNIEnv *env, jobject obj, jint len){
+	int ret = 512;
+	if(len>0)
+		ret = len;
+	char *cData = (char*)malloc(sizeof(char)*ret);
```

Comment:
Is this allocation freed by the JVM? Same question for Java_com_github_smallru8_driver_tuntap_TunTap_tuntap_1write.